Firestore/core/src/firebase/firestore/util/status.h

```diff
@@ -120,10 +120,28 @@ class ABSL_MUST_USE_RESULT Status {
 
  private:
   static const std::string& empty_string();
+
   struct State {
     State() = default;
     State(const State& other);
-    State(Error code, std::string msg);
+    State(Error code, std::string&& msg);
+
+    struct Deleter {
+      void operator()(const State* ptr) const;
+    };
+    // A `unique_ptr` with a custom deleter. If the pointer's value has been set
+    // to a special value (0x01) to indicate it is moved, invoking the custom
+    // deleter will be a no-op.
+    using StatePtr = std::unique_ptr<State, Deleter>;
+
+    static State* MovedFromIndicator() {
+      return reinterpret_cast<State*>(0x01);
+    }
+
+    template <typename... Args>
+    static StatePtr MakePtr(Args&&... args) {
+      return StatePtr(new State(std::move(args)...));
```

Comment:
This should be `std::forward<Args>`, not `std::move`.

The idea behind forwarding is to make sure the function to which the arguments are forwarded receives them _exactly_ as they were in the function that does forwarding. The more correct technical description is "preserving [value category](https://en.cppreference.com/w/cpp/language/value_category)", where "value category" basically means "is this an lvalue or an rvalue". Colloquially, it can be said that rvalues are temporaries and lvalues are non-temporaries.

In simpler terms, imagine we have two overloads of a function `Foo`, which only differ on whether they take the argument by const lvalue reference or an rvalue reference:
```cpp
void Foo(const std::string& str) { std::cout << "Lvalue overload\n"; }
void Foo(std::string&& str) { std::cout << "Rvalue overload\n"; }
```
Now we want to write a forwarding template function such that it forwards its argument to `Foo` _and_ invokes the correct overload:
```cpp
template <typename T>
void Fwd(T&& arg) {
  Foo(/*how do we pass arg to Foo here?*/);
}

std::string s;
Fwd(s); // Should invoke the lvalue overload, because `s` is not a temporary
Fwd(std::string{}); // Should invoke the rvalue overload, because the string is a temporary
```
(note: for simplicity, I make this a non-variadic template function. Whether the function is variadic or not is orthogonal to forwarding)

If we simply pass `arg` to `Foo`:
```cpp
template <typename T>
void Fwd(T&& arg) {
  Foo(arg);
}
```
the result would be that it _always_ chooses the lvalue overload, even when invoked with a temporary! Let's see how that happens. When invoked with a temporary, the type of `arg` within `Fwd` will be `std::string&&`, that is, it will be an rvalue reference -- which is correct. However, somewhat counterintuitively, an rvalue reference is by itself an lvalue -- that is, it _refers_ to an rvalue, but the reference itself is an lvalue. That is, when we use `arg` within the body of `Fwd`, it's actually an lvalue.

Unfortunately, applying `std::move` is also not a perfect solution:
```cpp
template <typename T>
void Fwd(T&& arg) {
  Foo(std::move(arg));
}
```
The problem is that `std::move` is unconditional. If previously `arg` was never moved, now the problem is that it is _always_ moved. Any arguments passed to `Fwd` will now invoke the rvalue overload:
```cpp
std::string s;
Fwd(s); // Invokes the rvalue overload
Fwd(std::string{}); // Invokes the rvalue overload
```
This is actually worse than before, because it means `s` may be moved from inside the implementation of `Foo`, even though the calling code never explicitly requested a move.

So we have one solution that never moves and one solution that always moves. This is where `std::forward` comes into play.

One way of looking at `std::forward` is that it conditionally moves. Another way to describe it is that it preserves the type of reference exactly as it was. If `arg` was an rvalue reference, `std::forward`ing `arg` will also produce an rvalue. If `arg` was an lvalue reference, `std::forward` will preserve its "lvalueness". This is exactly the behavior we want:
```cpp
template <typename T>
void Fwd(T&& arg) {
  Foo(std::forward<T>(arg));
}

std::string s;
Fwd(s); // Invokes the lvalue overload, correctly
Fwd(std::string{}); // Invokes the rvalue overload, correctly
```

See [live example](https://wandbox.org/permlink/T196w2N21ZPBtAYM). I'm happy to follow up on this if you have any questions (it's a complicated topic).